# Class 06-07 - More Functions, Modules and Objects  
**COMP130 - Introduction to Computing**  
**Dickinson College**  

### Names:

As usual, your name(s) should go here.

### Objects

__Q1:__ An *object* is made up of what two things?  Briefly define each.

1. state
 * An object's _state_ is the variables that define that object, along with the values of the variables.
2. methods
 * An object's _methods_ are the functions that change or make use of the state of the object.

__Q2:__ What is a *constructor*?  What is its job?

A constructor is a special method that creates (or *initializes*) a new object.

__Q3:__ Imagine some of the code behind Netflix.  What *state* might an object representing a movie need to contain?

* name --> "Miller's Crossing"
* year --> 1990
* director --> "Joel Coen"
* genre --> "gangster"
* star1 --> "Gabriel Byrne"
* star2 --> "Marcia Gay Harden"
* playback_speed --> "normal"
* current_scene --> 1

__Q4:__ What *methods* might an object representing a movie need to contain?

* play_movie( )
* pause_movie( )
* next_scene( )
* prev_ scene( )
* subtitle_lang( )
* mute( )

__Q5:__ For one method in __Q4__ explain how it would use or modify the state of a movie object.

The *next_scene( )* method would change the *current_scene* the movie is playing to the following scene in the movie.

### `Turtle` Objects

The documentation for the `turtle` module on the course home page and the `TurtleExample.ipynb` notebook from class (in the same folder as this notebook) will be useful in answering some of the following questions.

__Q6:__ Run the following code cell and then draw an *object state diagram* representing the state of the `Turtle` referred to by `meg` when the program has finished. Take a picture of your object state diagram and put it into the markdown field below.

In [1]:
import turtle
meg=turtle.Turtle()

* xcor --> 0
* ycor --> 0

__Q7:__ Run the following code cell. Update your *object state diagram* to match `meg`s new state.  Take a picture of your object state diagram and put it into the markdown field below.

In [2]:
meg.pencolor('blue')
meg.forward(50)
meg.left(90)
meg.forward(100)
meg.penup()

* xcor --> 50
* ycor --> 100
* heading --> 90
* pencolor --> 'blue'
* penup --> up

__Q8:__ Run the code cell below, update your state diagram to match `meg`'s new state.  Take a picture of your updated object state diagram and put it into the markdown field below.  Hint: Don't compute the state by hand... add a code cell.

In [3]:
meg.right(38)
meg.forward(82)

In [4]:
print(meg.xcor())
print(meg.ycor())
print(meg.heading())
print(meg.pencolor())
print(meg.penup())

100.48424097670397
164.61688179575123
52.0
blue
None


xcor --> 100.48424097670397
ycor --> 164.61688179575123
heading --> 52.0
pencolor --> blue
penup --> None

Note: The state of `meg.penup( )` is returned as _None_ because `penup( )` is a method that _does_ something (stops the turtle drawing) rather than a function that _returns_ some piece of data. (Notice that `penup()` takes no argument!)

See what happens when we try to get the __type__ of data that `penup( )` delivers:

In [5]:
type(meg.penup())

NoneType

__Q9:__ Use the *Reset and Clear Output* option on the *Kernel* menu to close the Turtle Window. Write a program that creates a `Turtle` and causes it to draw a rectangle that is not a square.  Be sure to import the `turtle` module into your program.

In [6]:
from turtle import *

sqTurt = Turtle()
sqTurt.pencolor('blue')
sqTurt.forward(100)
sqTurt.left(90)
sqTurt.forward(50)
sqTurt.left(90)
sqTurt.forward(100)
sqTurt.left(90)
sqTurt.forward(50)
sqTurt.penup()

__Q10:__ Use the *Reset and Clear Output* option on the *Kernel* menu to close the Turtle Window. Write a program that uses the `goto` method in the `Turtle` class to draw a right triangle. Be sure to import the turtle module into your program.

In [7]:
from turtle import *

triTurt = Turtle()
triTurt.pencolor('blue')
triTurt.forward(100)
triTurt.left(90)
triTurt.forward(50)
triTurt.left(150)
triTurt.goto(0,0)
triTurt.penup()

__Q11:__ What method in the `Turtle` class will draw part of a circle?

`turtle.circle( )` will draw part of a circle if the 2nd argument it receives is a number less than 360.

__Q12:__ What is the name and position of the argument to the method from __Q11__ that determines how much of the circle to draw?

The 2nd argument that `turtle.circle( )` takes is called _extent_.

__Q13:__ Use the *Reset and Clear Output* option on the *Kernel* menu to close the Turtle Window. Write a program that make a line drawing of [a half moon](https://pixabay.com/photos/moon-celestial-body-half-moon-927215/). Be sure to import the turtle module into your program.

In [8]:
from turtle import *

moonTurt = Turtle()
moonTurt.pencolor('green')
moonTurt.circle(100, 180)
moonTurt.left(90)
moonTurt.forward(200)

__Q14:__ Since we all know that [the moon is made of green cheese](https://en.wikipedia.org/wiki/The_Moon_is_made_of_green_cheese) modify your program in __Q13__ so that your half moon is *filled* in green. Hint: You'll need to use some additional `Turtle` methods that control *filling* shapes.

In [9]:
from turtle import *

moonTurt = Turtle()
moonTurt.pencolor('green')
moonTurt.fillcolor('green')
moonTurt.begin_fill()
moonTurt.circle(100, 180)
moonTurt.left(90)
moonTurt.forward(200)
moonTurt.end_fill()

__Q15:__ Use the *Reset and Clear Output* option on the *Kernel* menu to close the Turtle Window.  Write a program that draws a picture using turtle graphics. Be sure to import the turtle module into your program.

In [10]:
from turtle import *

sqTurt = Turtle()

sqTurt.pencolor('red')
sqTurt.fillcolor('red')
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)

sqTurt.forward(25)

sqTurt.pencolor('yellow')
sqTurt.fillcolor('yellow')
sqTurt.pendown()
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)

sqTurt.forward(25)

sqTurt.pencolor('red')
sqTurt.fillcolor('red')
sqTurt.pendown()
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)

sqTurt.goto(0, -25)

sqTurt.pencolor('yellow')
sqTurt.fillcolor('yellow')
sqTurt.pendown()
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)

sqTurt.forward(25)

sqTurt.pencolor('red')
sqTurt.fillcolor('red')
sqTurt.pendown()
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)

sqTurt.forward(25)

sqTurt.pencolor('yellow')
sqTurt.fillcolor('yellow')
sqTurt.pendown()
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)

sqTurt.goto(0, -50)

sqTurt.pencolor('red')
sqTurt.fillcolor('red')
sqTurt.pendown()
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)

sqTurt.forward(25)

sqTurt.pencolor('yellow')
sqTurt.fillcolor('yellow')
sqTurt.pendown()
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)

sqTurt.forward(25)

sqTurt.pencolor('red')
sqTurt.fillcolor('red')
sqTurt.pendown()
sqTurt.begin_fill()
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.left(90)
sqTurt.forward(25)
sqTurt.end_fill()
sqTurt.penup()
sqTurt.left(90)


If the length and inefficiency of the code above bothers you, then you are ready for Class08!

![Stop sign](stop.png) End of Class 06 material.

### Another Module and another Object

Included in the folder for this class is the file `bank.py`. This file contains a module named `bank` that defines a class named `Account`.  In the exercises below you will learn more about Python objects by creating and using `Account` objects.

__Q16:__ Write a statement that imports the `bank` module.

In [11]:
from bank import *

#### Constructors and Initialization

When we created new `Turtle` objects we used a statement like: `anh=turtle.Turtle()` to invoke the constructor and initialize the state of the new `Turtle` object.  Because every `Turtle` is exactly the same to begin with, the constructor does not need any additional information in order to initialize the state. In more CS terms, the constructor didn't require any *arguments*, thus the `()` after `turtle.Turtle` are empty.

The state of an `Account` object includes:
- `name` : the name of the account holder.
- `account_number`: a unique numeric identifier for the account.
- `balance` : the amount of money in the account.

Unlike `Turtle` objects which are all initially the same, each `Account` object will need a different `name` and a different `account_number`. To accomplish this the constructor for `Account` objects accepts arguments that provide the values indicating the name and account number.  So the constructor would be described as: 

- `Account(name, number)` : construct new `Account` with the name and account number.

Thus, to create a new `Account` you will use a statement like: 

- `savings=bank.Account('Grant', 1234)`.  

The arguments `'Grant'` and `1234` provide the information that the `Account` constructor needs to initialize the object. If you think about it, this really is no different than passing arguments to any other function. I.e. It is just like providing the argument `100` in `meg.forward(100)` or the `'Hello World!'` argument in `print('Hello World!')` provide information to the `forward` and `print` functions about what they are to do.

It is also important to note that the `Account` constructor sets the initial balance of every account to be 0.

__Q17:__ Write a statement that creates a new `Account` for yourself with the account number `98765`.

In [ ]:
new_acc = Account("Lev", 98765)

__Q18:__ Draw an object state diagram for the object you created in __Q17__ and include it in the markdown cell below.

Account - new_acc
* name --> "Lev"
* number --> 4321

All `Account` objects have the following methods:
- `get_name()` : returns the name of the account holder.
- `get_number()` : returns the account number.
- `get_balance()` : returns the current balance of the account.
- `deposit(amount)` : deposit the `amount` into the account.
- `withdraw(amount)` : withdraw the `amount` from the account.

__Q19:__ Write statements below that make a deposit into the `Account` you created in __Q18__ and then prints out the current balance of the account.

In [13]:
new_acc.deposit(1000000)
print("$" + str(new_acc.get_balance()))

$1000000


__Q20:__ What happens when you run the code cell from __Q19__ multiple times?

The Q19 code cell adds $1 million to newAcc each time it is run.

__Q21:__ Write statements below that make two different withdraws from the `Account` you created in __Q18__ and then prints out the current balance of the account.

In [14]:
new_acc.withdraw(50000)
new_acc.withdraw(5)
print("$" + str(new_acc.get_balance()))

$949995


__Q22:__ Write statements below that create another `Account` object.

In [15]:
acc_new = Account("Famine Relief", 4321)

__Q21:__ Bank customers often transfer money between their accounts.  Write a program below that prompts the user for an amount and then transfers that amount from the account created in __Q19__ to the account in __Q22__.

In [16]:
print(new_acc.get_name() + ", please help end hunger.")
donation = float(input("Please enter an amount to donate: "))
new_acc.withdraw(donation)
acc_new.deposit(donation)
print("Thank you!")
print(new_acc.get_name() + "'s account now has $" + str(new_acc.get_balance()) + " remaining.")
print(acc_new.get_name() + "'s account now has $" + str(acc_new.get_balance()) + " in it.")

Lev, please help end hunger.
Please enter an amount to donate: 500000
Thank you!
Lev's account now has $449995.0 remaining.
Famine Relief's account now has $500000.0 in it.


__Q22:__ There are also ways in which accounts might be configured to make automatic transfers.  For example, to support someone wishing to build up savings a bank might offer a service like automatically transferring 1/2 of a customer's checking account balance to their savings account at the end of each month.  

Write a program below that transfers 1/2 of the current balance from the account created in __Q19__ to the account in __Q22__.

In [17]:
print("Okay, " + new_acc.get_name() + ", now we really need funds.")
print("We know you're nice, so we're going to take half your money.")
donation = new_acc.get_balance() / 2
new_acc.withdraw(donation)
acc_new.deposit(donation)
print("Thank you!")
print(new_acc.get_name() + "'s account now has $" + str(new_acc.get_balance()) + " remaining.")
print(acc_new.get_name() + "'s account now has $" + str(acc_new.get_balance()) + " in it.")

Okay, Lev, now we really need funds.
We know you're nice, so we're going to take half your money.
Thank you!
Lev's account now has $224997.5 remaining.
Famine Relief's account now has $724997.5 in it.


### Fruitful and Fruitless (aka *void*) Methods/Functions

__Q23:__ Which of the methods in `Account` are *fruitful*?

* `get_name( )`
* `get_number( )`
* `get_balance( )`

__Q24:__ Which of the methods in `Account` are *fruitless* (or *void*)?

* `deposit( )`
* `withdraw( )`

As we know *fruitful functions* return values that we can save in a variable and/or use in other computations.  For example, as you did with `get_balance()` in __Q21__ and __Q22__ above.  However, *fruitless functions* do not return values and thus have no results that can be used in further computations.  Sometimes however, programmers make mistakes and assign a fruitless function to a variable.

__Q25:__ What value is assigned to the variable `hummm` by the following statements? 

In [18]:
my_acct=Account("Me", 3456)
hummm=my_acct.deposit(100)
print(hummm)

None


The value `None` is assigned to the variable `hummm` if you use the `deposit( )` method to assign it a value.  `None` is a special value that Python uses when there is no meaningful value associated with a variable.

__Q26__: What happens if you attempt to use `hummm` in another computation? E.g. add 10 to it? Or multiply it by 2?  Which of the three types of errors (*syntax*, *runtime*, *semantic*) is that?

In [19]:
hummm = hummm + 10
print(hummm)

TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'

Since `hummm` has no value (i.e. `None`), it cannot be added to an integer or anything else.

__Q27:__ Run the code in the code cell below a few times to understand what it does.  Then, in the code cell below that write a program that prints the distance between hieu and amy.

In [21]:
import turtle
import random

hieu=turtle.Turtle()
amy=turtle.Turtle()

hieu.pencolor('green')
hieu.forward(random.randint(50,150))
hieu.right(random.randint(30,120))
hieu.forward(random.randint(50,150))

amy.pencolor('yellow')
amy.left(90)
amy.forward(random.randint(50,150))
amy.left(random.randint(30,120))
amy.forward(random.randint(50,150))

In [22]:
dist_h_a = hieu.distance(amy.xcor(), amy.ycor())
print("The distance between turtles is " + str(dist_h_a) + ".")

The distance between turtles is 293.7885952459318.
